File: src/domain/task.py
"""Research task domain object."""
from __future__ import annotations

from abc import ABC
from abc import abstractmethod
from datetime import datetime
from typing import Final
from uuid import uuid4

from src.domain.result import Serializable
from src.domain.result import TaskResult


class Identifiable(ABC):
    """Object with unique identifier."""

    @abstractmethod
    def id(self) -> str:
        """Return unique identifier."""
        ...


class Queryable(ABC):
    """Object with research query."""

    @abstractmethod
    def query(self) -> str:
        """Return research query."""
        ...


class Served(ABC):
    """Object with service name."""

    @abstractmethod
    def service(self) -> str:
        """Return service name."""
        ...


class ResearchTask(Identifiable, Queryable, Served, Serializable):
    """Immutable research task with query and result."""

    def __init__(
        self,
        query: str,
        status: str,
        result: TaskResult | None,
        language: str = "русский",
        service: str = "parallel.ai",
        identifier: str | None = None,
        created: datetime | None = None,
        completed: datetime | None = None,
    ) -> None:
        """Initialize task with query, status, and optional result."""
        self._id: Final[str] = identifier or str(uuid4())
        self._query: Final[str] = query
        self._status: Final[str] = status
        self._result: Final[TaskResult | None] = result
        self._language: Final[str] = language
        self._service: Final[str] = service
        self._created: Final[datetime] = created or datetime.now()
        self._completed: Final[datetime | None] = completed

    def id(self) -> str:
        """Return task identifier."""
        return self._id

    def query(self) -> str:
        """Return research query."""
        return self._query

    def status(self) -> str:
        """Return task status."""
        return self._status

    def result(self) -> TaskResult | None:
        """Return task result if completed."""
        return self._result

    def language(self) -> str:
        """Return research language."""
        return self._language

    def service(self) -> str:
        """Return service name."""
        return self._service

    def created(self) -> datetime:
        """Return creation timestamp."""
        return self._created

    def completed(self) -> datetime | None:
        """Return completion timestamp."""
        return self._completed

    def complete(self, result: TaskResult) -> ResearchTask:
        """Return new task marked as completed with result."""
        return ResearchTask(
            query=self._query,
            status="completed",
            result=result,
            language=self._language,
            service=self._service,
            identifier=self._id,
            created=self._created,
            completed=datetime.now(),
        )

    def serialize(self) -> dict:
        """Return dictionary representation."""
        data = {
            "id": self._id,
            "query": self._query,
            "status": self._status,
            "language": self._language,
            "service": self._service,
            "created": self._created.isoformat(),
        }
        if self._completed:
            data["completed"] = self._completed.isoformat()
        if self._result:
            data["result"] = self._result.serialize()
        return data

    @classmethod
    def deserialize(cls, data: dict) -> ResearchTask:
        """Create ResearchTask from dictionary."""
        result = None
        if "result" in data:
            result = TaskResult.deserialize(data["result"])
        completed = None
        if "completed" in data:
            completed = datetime.fromisoformat(data["completed"])
        return cls(
            query=data["query"],
            status=data["status"],
            result=result,
            language=data.get("language", "русский"),
            service=data.get("service") or "parallel.ai",
            identifier=data["id"],
            created=datetime.fromisoformat(data["created"]),
            completed=completed,
        )

File: src/research/domain/task.clj
(ns research.domain.task
  (:refer-clojure :exclude [format])
  (:require [clojure.string :as str]
            [research.domain.result :as result])
  (:import (java.time LocalDateTime)
           (java.time.format DateTimeFormatter)
           (java.util Optional UUID)))

(defprotocol Tasked
  "Object representing a research task."
  (id [item] "Return task identifier.")
  (query [item] "Return research query.")
  (status [item] "Return task status.")
  (report [item] "Return task result object.")
  (language [item] "Return task language.")
  (provider [item] "Return task provider.")
  (created [item] "Return creation time.")
  (completed [item] "Return completion time.")
  (finish [item value] "Return task marked as completed.")
  (data [item] "Return map representation."))

(defn now
  "Return current local datetime."
  []
  (LocalDateTime/now))

(defn parse
  "Parse ISO datetime string into LocalDateTime."
  [text]
  (LocalDateTime/parse text))

(defn format
  "Format LocalDateTime into ISO string."
  [time]
  (.format time DateTimeFormatter/ISO_LOCAL_DATE_TIME))

(defrecord ResearchRun [id brief data result]
  Tasked
  (id [_] id)
  (query [_]
    (let [text (str (or (:text brief) ""))
          topic (str (or (:topic brief) ""))
          items (or (:items brief) [])
          items (vec (remove str/blank? items))
          rows (map-indexed (fn [idx item] (str (inc idx) ". " item)) items)
          tail (str/join "\n" rows)
          note (cond
                 (not (str/blank? text)) text
                 (and (not (str/blank? topic)) (seq rows))
                 (str topic "\n\nResearch:\n" tail)
                 (seq rows) (str "Research:\n" tail)
                 :else topic)]
      note))
  (status [_] (:status data))
  (report [_] result)
  (language [_] (:language data))
  (provider [_] (:service data))
  (created [_] (:created data))
  (completed [_] (:completed data))
  (finish [_ value]
    (->ResearchRun
     id
     brief
     (assoc data :status "completed" :completed (Optional/of (now)))
     value))
  (data [_] (let [base {:id id
                        :status (:status data)
                        :language (:language data)
                        :service (:service data)
                        :processor (:processor data)
                        :brief brief
                        :created (format (:created data))}
                  done (:completed data)
                  ready (if (.isPresent done)
                          (assoc base :completed (format (.get done)))
                          base)
                  proc (:processor data)
                  ready (if (str/blank? (str proc))
                          (dissoc ready :processor)
                          ready)]
              ready)))

(defn task
  "Create task from map."
  [item]
  (let [text (or (:language item) "русский")
        name (or (:service item) "parallel.ai")
        time (parse (:created item))
        done (if (:completed item)
               (Optional/of (parse (:completed item)))
               (Optional/empty))
        entry (:brief item)
        query (or (:text entry) (:query item) "")
        rows (str/split-lines (str query))
        label "Research:"
        spot (first (keep-indexed
                     (fn [idx line]
                       (when (= label (str/trim line)) idx))
                     rows))
        edge (first (keep-indexed
                     (fn [idx line]
                       (let [line (str/trim line)
                             hit (re-find
                                  #"^(?:\d+\s*[\.)]|\d+|[*+-])\s+"
                                  line)]
                         (when hit idx)))
                     rows))
        cut (if (some? spot) spot edge)
        head (vec (if (some? cut) (take cut rows) rows))
        tail (if (some? cut)
               (drop (if (some? spot) (inc spot) cut) rows)
               [])
        list (loop [list [] chunk tail]
               (if (seq chunk)
                 (let [line (str/trim (first chunk))]
                   (if (str/blank? line)
                     (recur list (rest chunk))
                     (let [num (re-find #"^\d+\s*[\.)]\s+(.+)$" line)
                           raw (re-find #"^\d+\s+(.+)$" line)
                           bul (re-find #"^[*+-]\s+(.+)$" line)
                           part (cond
                                  num (second num)
                                  raw (second raw)
                                  bul (second bul)
                                  :else nil)
                           part (if (and part (not (str/blank? part)))
                                  (str/trim part)
                                  nil)
                           list (if part
                                  (conj list part)
                                  (if (seq list)
                                    (conj (vec (butlast list))
                                          (str (last list) " " line))
                                    (conj list line)))]
                       (recur list (rest chunk)))))
                 list))
        list (vec (remove str/blank? (map str/trim list)))
        top (reduce
             (fn [text line]
               (if (str/blank? (str/trim line)) text (str/trim line)))
             ""
             head)
        topic (or (:topic entry) top "")
        items (vec (or (:items entry) list []))
        brief {:text query
               :topic topic
               :items items}
        data {:status (:status item)
              :language text
              :service name
              :processor (:processor item)
              :created time
              :completed done}
        raw (:result item)
        value (result/result raw)
        code (or (:id item) (str (UUID/randomUUID)))]
    (->ResearchRun code brief data value)))
